src/free_functions.rs

```diff
@@ -87,22 +87,16 @@ pub fn aview1<A>(xs: &[A]) -> ArrayView1<'_, A> {
 
 /// Create a two-dimensional array view with elements borrowing `xs`.
 ///
-/// **Panics** if the product of non-zero axis lengths overflows `isize`. (This
-/// can only occur when `V` is zero-sized.)
+/// **Panics** if the product of non-zero axis lengths overflows `isize`.
 pub fn aview2<A, V: FixedInitializer<Elem = A>>(xs: &[V]) -> ArrayView2<'_, A> {
     let cols = V::len();
     let rows = xs.len();
     let dim = Ix2(rows, cols);
-    if size_of::<V>() == 0 {
```

Comment:
Previously we only needed to check if `row * col` is above `isize::MAX` for arrays of ZSTs, since `FixedInitializer` was only implemented for arrays of <16 and slices cannot contain more than `isize::MAX` number of bytes. Now that `FixedInitializer` is implemented for arrays of any length, `row * col` can be above `isize::MAX` even for non-ZSTs, so we perform the check unconditionally.

Reply:
I don't agree because we know that if there is a materialized "real array" value, then it is allowed as long as the element type is non-ZST.

Reply:
so you're saying that, since slices can only cover up to `isize::MAX` bytes, that the slice can never contain more than `isize::MAX` non-ZST elements? I get what you're saying.

Reply:
Rust says that its allocated object may not be larger than isize. Since we take a object that already exists here, we don't need to check its size. :)

Reply:
What if we're creating views from stack-allocated arrays? Does the allocation limit still apply?

Reply:
I'd find the RalfJung quote for you but I don't know where I found it. Yes, allocated object refers to abstract allocations, i.e what in the C standard would be called objects, not in the std::alloc sense :)

Reply:
So should we make the bounds check conditional or leave it as it is? I think we can "play it safe" just in case there's some corner case we don't know about. The extra check isn't a concern because `aview2` isn't performance-critical.

Reply:
I would think that it would be possible to construct a value of type `&[[u8; usize::MAX]]` if the slice was zero-length (since the "allocated object" would be zero-sized), but I can't figure out any way to do it. Neither of these work:

```rust
let s: &[[u8; usize::MAX]] = &[];
let s: &'static [[u8; std::usize::MAX]] =
    unsafe { std::slice::from_raw_parts(std::ptr::NonNull::dangling().as_ptr(), 0) };
```

The Rust 1.51 compiler says when building (not when checking): "error: values of the type `[u8; 18446744073709551615]` are too big for the current architecture".

Reply:
Please keep the logic for the check as it was before the PR

Reply:
[Here's a reference for "allocated objects have a maximum size of `isize::MAX`"](https://doc.rust-lang.org/stable/reference/types/numeric.html#machine-dependent-integer-types):

> The `isize` type is a signed integer type with the same number of bits as the platform's pointer type. The theoretical upper bound on object and array size is the maximum `isize` value. This ensures that `isize` can be used to calculate differences between pointers into an object or array and can address every byte within an object along with one byte past the end.